btree/btree.c

```diff
@@ -3547,46 +3466,96 @@ static int64_t btree_put_kv_tick(struct m0_sm_op *smop)
 				node_unlock(lev->l_node);
 				if (oi->i_key_found)
 					return P_LOCK;
-				return P_ALLOC;
+				/**
+				 * Initialized i_alloc_lev to level of leaf
+				 * node.
+				 */
+				oi->i_alloc_lev = oi->i_used;
+				return P_ALLOC_REQUIRE;
 			}
 		} else {
 			node_op_fini(&oi->i_nop);
 			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
 		}
-	case P_ALLOC: {
-		bool alloc = false;
+	case P_ALLOC_REQUIRE:{
 		do {
-			lev = &oi->i_level[oi->i_used];
-			/**
-			 * Validate node to dertmine if lev->l_node is still
-			 * exists.
-			 */
+			lev = &oi->i_level[oi->i_alloc_lev];
 			if (!node_isvalid(lev->l_node)) {
-				oi->i_used = oi->i_height - 1;
 				return m0_sm_op_sub(&bop->bo_op, P_CLEANUP,
 						    P_SETUP);
 			}
 			if (!node_isoverflow(lev->l_node))
 				break;
-			if (oi->i_used == 0) {
-				if (lev->l_alloc == NULL ||
-				    oi->i_extra_node == NULL)
-					alloc = true;
-				break;
-			} else if (lev->l_alloc == NULL) {
-				alloc = true;
-				break;
-			}
-
-			oi->i_used--;
-		} while (1);
+			if (lev->l_alloc == NULL || (oi->i_alloc_lev == 0 &&
+			    oi->i_extra_node == NULL)) {
+				/**
+				 * Depending on the level of node, shift can be
+				 * updated.
+				 */
+				int ksize = node_keysize(lev->l_node);
+				int vsize = node_valsize(lev->l_node);
+				int shift = node_shift(lev->l_node);
+				oi->i_nop.no_opc = NOP_ALLOC;
+				return node_alloc(&oi->i_nop, tree,
+						  shift,lev->l_node->n_type,
+						  ksize, vsize, lock_acquired,
+						  bop->bo_tx, P_ALLOC_STORE);
 
-		if (alloc)
-			return btree_put_alloc_phase(bop);
-		/* Reset oi->i_used */
-		oi->i_used = oi->i_height - 1;
+			}
+			oi->i_alloc_lev--;
+		} while (oi->i_alloc_lev >= 0);
 		return P_LOCK;
 	}
+	case P_ALLOC_STORE: {
+		lev = &oi->i_level[oi->i_alloc_lev];
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			if (oi->i_alloc_lev == 0) {
+				/**
+				 * If we are at root node and if there is
+				 * possibility of overflow at root node,
+				 * allocate two nodes for l_alloc and
+				 * oi->extra_node, as both nodes will
+				 * be used in case of root overflow.
+				 */
+				if (lev->l_alloc == NULL) {
+					int ksize;
+					int vsize;
+					int shift;
+					lev->l_alloc = oi->i_nop.no_node;
+
+					if (!node_isvalid(lev->l_node)) {
+						return m0_sm_op_sub(&bop->bo_op,
+								    P_CLEANUP,
+								    P_SETUP);
+					}
+					ksize = node_keysize(lev->l_node);
+					vsize = node_valsize(lev->l_node);
+					shift = node_shift(lev->l_node);
+					oi->i_nop.no_opc = NOP_ALLOC;
+					return node_alloc(&oi->i_nop, tree,
+							  shift,
+							  lev->l_node->n_type,
+							  ksize, vsize,
+							  lock_acquired,
+							  bop->bo_tx,
+							  P_ALLOC_STORE);
+
+				} else if (oi->i_extra_node == NULL){
+					oi->i_extra_node = oi->i_nop.no_node;
+					return P_LOCK;
+				} else
+					M0_ASSERT(0);
```

Comment:
For root level , it will come here only if `l_alloc` is Null of `i_extra_node` is Null. I think we can eliminate this else check

Reply:
Yes else check can be removed, but it is added intentionally to ensure that we should not reach this step.